Sources/NIOCore/NIOOnLoopSendable.swift

```diff
@@ -0,0 +1,78 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2023 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+public struct NIOOnLoopSendable<Value>: @unchecked Sendable {
+    @usableFromInline
+    let _eventLoop: EventLoop
+
+    @usableFromInline
+    var _value: Value
+
+    @inlinable
+    public init(_ value: Value, eventLoop: EventLoop) {
+        eventLoop.preconditionInEventLoop()
+        self._eventLoop = eventLoop
+        self._value = value
+    }
+
+    @inlinable
+    public var value: Value {
+        get {
+            self._eventLoop.preconditionInEventLoop()
+            return self._value
+        }
+        set {
+            self._eventLoop.preconditionInEventLoop()
+            self._value = newValue
+        }
+    }
+}
+
+public class NIOOnLoopSendableBox<Value>: @unchecked Sendable {
+    @usableFromInline
+    let _eventLoop: EventLoop
+
+    @usableFromInline
+    var _value: Value
+
+    @inlinable
+    public init(_ value: Value, eventLoop: EventLoop) {
+        eventLoop.preconditionInEventLoop()
```

Comment:
Isn't this precondition overly restrictive? Why only allow `Optinal`s to be initialised to `.none` without being on the given `eventLoop` but nothing else?

Reply:
Because the `none` value is definitely safe to send, as it doesn't contain the non-Sendable value.